Tempus:  Add combined, staggered, and pseudotransient FSA methods

etphipp opened the pull request:
This pull-request adds integrators supporting the combined, staggered, and
pseudotransient forward sensitivity analysis methods where the
sensitivity equations are solved alongside the forward equations.  The
different methods refer to how the solves between states/sensitivities
are partitioned:

  * combined:  solved together (best for explicit)
  * staggered: solve state equations and then sensitivity equations for
    each time step (best for implicit)
  * pseudotransient: do full state equation integration followed by full
    sensitivity equation integration (for when time integrator is used
    as a steady-state solver)

In all cases there is an option, "Use DfDp as Tangent", that controls
whether the DfDp out-arg is interpreted as the tangent df/dp +
df/dx*dx/dp + df/dxdot*dxdot/dp, or not (which is not regular model
evaluator behavior).  If the model supports it, it is a much simpler and
more efficient way to assemble the sensitivity equations.

It also supports options to reuse the linear solver from the state
equations when solving the sensitivity equations.  This saves Jacobian
assembly time and preconditioner generation time.  Supporting it
required one small change to the NOX-Thyra interface to not throw when
accessing the preconditioner RCP if it is null (because there is no way
to know beforehand if it is null or not until you access it).

Tests were added for Backward Euler, Explicit RK, and BDF2 time
integration methods for all of the above methods and options.

@trilinos/tempus @trilinos/nox 

## Checklist
- [ ] My commit messages mention the appropriate GitHub issue numbers.
- [ x] My code follows the code style of the affected package(s).
- [ x] My change requires a change to the documentation.
- [x ] I have updated the documentation accordingly.
- [ ] I have read the [code contribution guidelines](../blob/master/CONTRIBUTING.md) for this project.
- [ x] I have added tests to cover my changes.
- [x ] All new and existing tests passed.
- [x ] No new compiler warnings were introduced.
- [ ] These changes break backwards compatibility.

ccober6 commented:
This looks great!  Sorry it took me some time, but I wanted to look it over and try to understand some of the pieces.  So unless @rppawlo  has additional comments, I would say let's merge it!

ibaned commented on packages/tempus/test/BDF2/Tempus_BDF2Test.cpp:
```diff
@@ -191,6 +195,251 @@ TEUCHOS_UNIT_TEST(BDF2, SinCos)
+    // Plot sample solution and exact solution
+    if (comm->getRank() == 0 && n == nTimeStepSizes-1) {
+      typedef Thyra::DefaultMultiVectorProductVector<double> DMVPV;
+
```
This broke pull request testing for a later pull request, see [here](/ascldap/users/daibane/src/alexa/src/LowRmPotentialSolve.hpp)

ibaned commented:
@ccober6 @etphipp please take a look at [this](https://github.com/trilinos/Trilinos/pull/2002#issuecomment-345381902), I think these changes are causing compile errors.

bartlettroscoe commented:
It seems likely that this merge broke the CI build as shown here:

* https://testing.sandia.gov/cdash/index.php?project=Trilinos&parentid=3225262

Here was the push log:

```
commit d18aea5cb4b90a5f96600e4a635db2f44e3b72db                                                                                                                                                                    
Merge: aeff34d e8e6d67                                                                                                                                                                                             
Author:     Curtis Ober <ccober@sandia.gov>                                                                                                                                                                        
AuthorDate: Fri Nov 17 14:12:22 2017 -0700                                                                                                                                                                         
Commit:     GitHub <noreply@github.com>                                                                                                                                                                            
CommitDate: Fri Nov 17 14:12:22 2017 -0700                                                                                                                                                                         
                                                                                                                                                                                                                   
    Merge pull request #1990 from trilinos/tempus_fsa                                                                                                                                                              
                                                                                                                                                                                                                   
    Tempus:  Add combined, staggered, and pseudotransient FSA methods                                                                                                                                              
                                                                                                                                                                                                                   
Commits pushed:                                                                                                                                                                                                    
d18aea5 Merge pull request #1990 from trilinos/tempus_fsa                                                                                                                                                          
e8e6d67 Tempus:  Add combined, staggered, and pseudotransient FSA methods                 
```

I will go ahead and back out this merge commit d18aea5 and push with checkin-test-sems.sh

bartlettroscoe commented:
I backed out the commit d18aea5 as the revert commit 4b51a54.    This has cleaned up the CI build as shown at:

* https://testing.sandia.gov/cdash/index.php?project=Trilinos&parentid=3226533

@etphipp and @ccober6,

My advice would be to rebase this topic branch on top of 'develop', run `./checkin-test-sems.sh --do-all` locally to show the build failures, and then fix the problems.  Once the problems are fixed, then run `./checkin-test-sems.sh --do-all --push` to push.

jwillenbring commented:
I will also reset the pull request testing to go back this far. That way if you just want to update the PR to fix the issues, it can run through the pull request testing prior to another merger attempt.

etphipp commented:
Sorry, I should have made it clear when posting the PR that it hadn't gone through SEMS-env testing yet, and that should happen before merging.  

Jim, I will probably merge it with the checkin test script anyway because I need to build it with the SEMS environment to understand why the compile is failing.

etphipp commented:
I fixed the build errors and pushed via the test script.  Some missing Thyra includes were needed.  I also split the sensitivity tests off into their own executables so that the tests don’t time out in debug builds.
